File: ultima version/scripts/verify-production-health.mjs
import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { Script } from 'node:vm';

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');

const htmlPages = [
    'index.html',
    'clases.html',
    'tarifas.html',
    'profile.html',
    'maestros.html',
    'success.html',
    'cancel.html',
    'politica-privacidad.html'
];

const jsFiles = [
    'public-calendar.js',
    'i18n.js',
    'teacher-profiles.js',
    'facilities-carousel.js'
];

let totalChecks = 0;
let passedChecks = 0;
const errors = [];

function check(name, pass, failureDetail = '') {
    totalChecks++;
    if (pass) {
        passedChecks++;
        console.log(`  ✅ [PASS] ${name}`);
    } else {
        errors.push({ name, detail: failureDetail });
        console.error(`  ❌ [FAIL] ${name}: ${failureDetail}`);
    }
}

async function run() {
    console.log('\n======================================================');
    console.log('🛡️  GEN YOGA PRODUCTION PRE-DEPLOY HEALTH CHECK');
    console.log('======================================================\n');

    // --- 1. SINTAXIS JAVASCRIPT EN HTML Y JS ---
    console.log('🔍 1. Comprobacion de sintaxis JavaScript y parsing VM...');
    for (const file of jsFiles) {
        try {
            const content = await readFile(path.join(root, file), 'utf8');
            new Script(content, { filename: file });
            check(`Sintaxis JS en ${file}`, true);
        } catch (e) {
            check(`Sintaxis JS en ${file}`, false, e.message);
        }
    }

    for (const page of htmlPages) {
        try {
            const html = await readFile(path.join(root, page), 'utf8');
            const scriptMatches = [...html.matchAll(/<script(?:\s+type="(?:text\/javascript|application\/javascript)")?[^>]*>([\s\S]*?)<\/script>/gi)];
            let scriptIndex = 1;
            let pageOk = true;
            for (const match of scriptMatches) {
                const code = match[1].trim();
                if (!code || match[0].includes('src=')) continue;
                try {
                    new Script(code, { filename: `${page}_script_${scriptIndex}` });
                } catch (err) {
                    pageOk = false;
                    check(`Sintaxis script #${scriptIndex} en ${page}`, false, err.message);
                }
                scriptIndex++;
            }
            if (pageOk) {
                check(`Sintaxis scripts en ${page}`, true);
            }
        } catch (e) {
            check(`Lectura de ${page}`, false, e.message);
        }
    }

    // --- 2. VERIFICACIÓN DE CONTROL DE ASISTENCIAS Y ALUMNOS (PROFILE.HTML) ---
    console.log('\n🔍 2. Control de asistencias y gestion de alumnos (profile.html)...');
    const profileHtml = await readFile(path.join(root, 'profile.html'), 'utf8');
    
    check('Funcion cargarAsistenciasPorClase presente', profileHtml.includes('async function cargarAsistenciasPorClase()'));
    check('Fallback de carga de clases en asistencias', profileHtml.includes("from('clases').select('*')") && profileHtml.includes("from('profesionales').select('*')"));
    check('Enriquecimiento de perfiles de alumnos desde reservas', profileHtml.includes('allAsistenciasPerfilesMap[r.user_id]'));
    check('Gestion de profesores en asistencias (esClaseDelProfesionalActual)', profileHtml.includes('function esClaseDelProfesionalActual(clase)'));
    check('Borrado seguro con reembolso y fallback de clases', profileHtml.includes('borrarClase(id)') && (profileHtml.includes('admin_eliminar_clase') || profileHtml.includes('saldo_gratis_descontado')));

    // --- 3. REGLA DE BONOS Y SESIONES GRATUITAS VS REGULARES ---
    check('Reconocimiento de sesiones abiertas/introductorias (esClaseElegibleBonoGratis)', profileHtml.includes('function esClaseElegibleBonoGratis(c)'));

    // --- 4. CALENDARIO PÚBLICO (PUBLIC-CALENDAR.JS) ---
    console.log('\n🔍 4. Calendario publico y clases de bienvenida (public-calendar.js)...');
    const calJs = await readFile(path.join(root, 'public-calendar.js'), 'utf8');
    check('Temporada activa desde 2026-08-24', calJs.includes("SEASON_START_WEEK = '2026-08-24'"));
    check('Inclusion de sesiones abiertas/introductorias en modo clases', calJs.includes('nombre.ilike.%introductor%,nombre.ilike.%abierta%'));
    check('No clasifica erroneamente clases de bienvenida como talleres de pago', calJs.includes('const isIntroOrOpen = /introductor|bienvenida|abierta|gratis|prueba/i.test(rawName);'));

    // --- 5. TARIFAS Y OFERTAS (TARIFAS.HTML) ---
    console.log('\n🔍 5. Estructura de tarifas y ofertas (tarifas.html)...');
    const tarifasHtml = await readFile(path.join(root, 'tarifas.html'), 'utf8');
    check('Oferta Única: Bono de Bienvenida (0 €) presente', (tarifasHtml.includes('rates_welcome_title') || tarifasHtml.includes('Bono de Bienvenida')) && tarifasHtml.includes('0 €'));
    check('Inactivadas ofertas anteriores en sección ofertas', !tarifasHtml.includes('rates_intro_badge') && !tarifasHtml.includes('rates_companion_eyebrow'));
    check('Pestanas de filtrado (Ofertas, Clases, Consultas, Talleres)', tarifasHtml.includes('tab-ofertas') && tarifasHtml.includes('tab-yoga') && tarifasHtml.includes('tab-psicologia') && tarifasHtml.includes('tab-talleres'));

    // --- 6. CANALES DE CONTACTO Y WHATSAPP ---
    console.log('\n🔍 6. Iconos de contacto y WhatsApp...');
    check('WhatsApp configurado en index.html', (await readFile(path.join(root, 'index.html'), 'utf8')).includes('https://wa.me/34624435679'));
    check('WhatsApp configurado en clases.html', (await readFile(path.join(root, 'clases.html'), 'utf8')).includes('https://wa.me/34624435679'));
    check('WhatsApp configurado en tarifas.html', tarifasHtml.includes('https://wa.me/34624435679'));

    // --- 7. CONSISTENCIA DE VERSIONES ---
    console.log('\n🔍 7. Consistencia de version...');
    const pkg = JSON.parse(await readFile(path.join(root, 'package.json'), 'utf8'));
    const currentVer = pkg.version;
    const shortVer = currentVer.split('.').slice(0, 2).join('.');
    check(`Version base en package.json (${currentVer})`, Boolean(currentVer));
    
    for (const page of htmlPages) {
        const content = await readFile(path.join(root, page), 'utf8');
        check(`Version v${shortVer} en ${page}`, content.includes(`?v=${shortVer}`) || content.includes(`v${shortVer}`));
    }

    console.log('\n======================================================');
    if (errors.length === 0) {
        console.log(`🎉 ¡TODOS LOS CONTROLES HAN PASADO CON ÉXITO! (${passedChecks}/${totalChecks})`);
        console.log('🚀 El proyecto esta 100% verificado y listo para produccion sin errores bloqueantes.');
        console.log('======================================================\n');
        process.exit(0);
    } else {
        console.error(`❌ SE ENCONTRARON ${errors.length} ERRORES BLOQUEANTES:`);
        errors.forEach(e => console.error(`   - ${e.name}: ${e.detail}`));
        console.error('======================================================\n');
        process.exit(1);
    }
}

run().catch(err => {
    console.error('Fatal test error:', err);
    process.exit(1);
});

File: ultima version/scripts/verify-production-health.py
#!/usr/bin/env python3
import glob
import json
import os
import re
import sys

def run_checks():
    print("\n======================================================")
    print("🛡️  GEN YOGA PRODUCTION PRE-DEPLOY HEALTH CHECK")
    print("======================================================\n")
    
    errors = []
    total = 0
    passed = 0
    
    def check(name, condition, detail=""):
        nonlocal total, passed
        total += 1
        if condition:
            passed += 1
            print(f"  ✅ [PASS] {name}")
        else:
            errors.append((name, detail))
            print(f"  ❌ [FAIL] {name}: {detail}")
            
    # 1. HTML Pages Readability & JavaScript Balance
    html_files = sorted(glob.glob("*.html"))
    for h in html_files:
        with open(h, "r", encoding="utf-8") as f:
            content = f.read()
        check(f"Archivo HTML {h} existe y no está vacío", len(content) > 100)
        opens = len(re.findall(r"<script\b", content, re.IGNORECASE))
        closes = len(re.findall(r"</script>", content, re.IGNORECASE))
        check(f"Etiquetas script balanceadas en {h} ({opens} open, {closes} close)", opens == closes)

    # 2. Profile.html Critical Functions
    with open("profile.html", "r", encoding="utf-8") as f:
        p_html = f.read()
    check("cargarAsistenciasPorClase presente", "async function cargarAsistenciasPorClase()" in p_html)
    check("Fallback de carga en asistencias presente", "from('clases').select('*')" in p_html and "from('profesionales').select('*')" in p_html)
    check("borrarClase presente con reembolso", "async function borrarClase(id)" in p_html and ("admin_eliminar_clase" in p_html or "saldo_gratis_descontado" in p_html))
    check("reservarClase con regla canónica (gratis vs regular)", "esSesionGratuita" in p_html and ("tieneBonoBienvenida" in p_html or "tieneBonoGratis" in p_html))

    # 3. Public Calendar Critical Rules
    with open("public-calendar.js", "r", encoding="utf-8") as f:
        cal_js = f.read()
    check("Temporada inicia en 2026-08-24", "SEASON_START_WEEK = '2026-08-24'" in cal_js)
    check("Inclusión de sesiones abiertas/introductorias en query", "nombre.ilike.%introductor%,nombre.ilike.%abierta%" in cal_js)
    check("Clases introductorias no son marcadas como talleres especiales", "const isIntroOrOpen = /introductor|bienvenida|abierta|gratis|prueba/i.test(rawName);" in cal_js)

    # 4. Tarifas Structure
    with open("tarifas.html", "r", encoding="utf-8") as f:
        tar_html = f.read()
    check("Ficha Bono Bienvenida presente", "rates_welcome_title" in tar_html or "Bono de Bienvenida" in tar_html)
    check("Ficha Sesiones Introductorias presente", "rates_intro_title" in tar_html or "Sesiones Introductorias" in tar_html)
    check("Ficha Yoga en Compañía presente", "rates_companion_title" in tar_html or "Yoga en compañ" in tar_html)
    check("4 pestañas (Ofertas, Clases, Consultas, Talleres)", "tab-ofertas" in tar_html and "tab-yoga" in tar_html and "tab-talleres" in tar_html)

    # 5. WhatsApp Contact Links
    with open("index.html", "r", encoding="utf-8") as f:
        idx_html = f.read()
    with open("clases.html", "r", encoding="utf-8") as f:
        cls_html = f.read()
    check("WhatsApp configurado en index.html", "https://wa.me/34624435679" in idx_html)
    check("WhatsApp configurado en clases.html", "https://wa.me/34624435679" in cls_html)
    check("WhatsApp configurado en tarifas.html", "https://wa.me/34624435679" in tar_html)

    # 6. Versión coherente
    with open("package.json", "r", encoding="utf-8") as f:
        pkg = json.load(f)
    v_full = pkg.get("version", "")
    v_short = ".".join(v_full.split(".")[:2])
    check(f"Versión base en package.json ({v_full})", bool(v_full))
    for h in html_files:
        with open(h, "r", encoding="utf-8") as f:
            c = f.read()
        check(f"Versión v{v_short} en {h}", f"?v={v_short}" in c or f"v{v_short}" in c)

    print("\n======================================================")
    if not errors:
        print(f"🎉 ¡TODOS LOS CONTROLES HAN PASADO CON ÉXITO! ({passed}/{total})")
        print("🚀 El proyecto está 100% verificado y listo para producción sin errores bloqueantes.")
        print("======================================================\n")
        return 0
    else:
        print(f"❌ SE ENCONTRARON {len(errors)} ERRORES BLOQUEANTES:")
        for name, detail in errors:
            print(f"   - {name}: {detail}")
        print("======================================================\n")
        return 1

if __name__ == "__main__":
    sys.exit(run_checks())
